devicetree.h: switch to DT_IRQ_INTC_BY_IDX for L2 and L3 INTID encodings

LaurentiuM1234 opened the pull request:
Using `DT_IRQ_INTC()` to fetch the interrupt controller associated with a node works well for nodes which consume interrupts from a single aggregator. However, when specifying multiple (and different) interrupt aggregators via the `interrupts-extended` property, the L2 and L3 interrupts will no longer be encoded properly. This is because `DT_IRQ_INTC(node_id)` uses `DT_IRQ_INTC_BY_IDX(node_id, 0)` so all the interrupts will use the first aggregator as their parent. To fix this, switch from using `DT_IRQ_INTC()` to `DT_IRQ_INTC_BY_IDX()`.

This can be tested using configurations such as the following:

```
intc1: interrupt-controller@cafebabe {
    /* some more properties here */
    interrupt-controller;
    #interrupt-cells = <1>;
};

intc2: interrupt-controller@deadbeef {
     /* some more properties here */
    interrupt-controller;
    interrupts-extended = <&intc1 14>;
};

intc3: interruptc-controller@babebeef {
     /* some more properties here */
    interrupt-controller;
    interrupts-extended = <&intc1 15>;
};

consumer_node: ip@cafecafe {
    /* some more properties here */
    interrupts-extended = <&intc2 16>, <&intc3 15>;
};

```

Without this patch the interrupts from `consumer_node` will be encoded as follows: `14 | ((16 + 1) << 8)` and `14 | ((15 + 1) << 8)`

Fixes #69820

LaurentiuM1234 committed:
devicetree.h: switch to DT_IRQ_INTC_BY_IDX for L2 and L3 INTID encodings

Using `DT_IRQ_INTC()` to fetch the interrupt controller associated
with a node works well for nodes which consume interrupts from a
single aggregator. However, when specifying multiple (and different)
interrupt aggregators via the `interrupts-extended` property,
the L2 and L3 interrupts will no longer be encoded properly. This
is because `DT_IRQ_INTC(node_id)` uses `DT_IRQ_INTC_BY_IDX(node_id, 0)`
so all the interrupts will use the first aggregator as their parent.
To fix this, switch from using `DT_IRQ_INTC()` to `DT_IRQ_INTC_BY_IDX()`.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

ycsin commented:
will take a look soon, but is it possible to extend the existing devicetree test for this?

LaurentiuM1234 committed:
tests: devicetree: check L2 interrupt encoding

Add some assert statements meant to check if L2 interrupts
are encoded right when dealing with nodes that consume interrupts
from multiple aggregators. For this to work, also add another
interrupt controller node which extends a different L1 interrupt
from `test_intc`.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 commented:
> will take a look soon, but is it possible to extend the existing devicetree test for this? i.e. with the DT configuration in the OP

created https://github.com/zephyrproject-rtos/zephyr/pull/69711

ycsin commented on tests/lib/devicetree/api/app.overlay:
```diff
@@ -440,6 +440,19 @@
 			interrupt-parent = <&test_cpu_intc>;
 		};
 
+		/* same as `test_intc` but extends a different L1 interrupt.
```
should be able to reuse existing nodes, see #69728

ycsin replied:
nvm, let's no go there for now, to keep this simple

ycsin reviewed (commented):
LGTM, I think this is a bug fix and should be backported, ~~just one comment about the test, which will be unblocked by #69728~~ nvm about the 2nd part of the comment

ycsin commented on tests/lib/devicetree/api/app.overlay:
```diff
@@ -440,6 +440,19 @@
+		 * Required for testing if interrupts are encoded properly for
+		 * nodes consumming interrupts from different aggregators.
+		 */
+		test_intc2: interrupt-controller@cafebabe {
```
interesting address, it works but I would suggest that we simply continue from address above (`bbbbdccc`)

LaurentiuM1234 replied:
fixed

ycsin commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -3201,6 +3201,17 @@ ZTEST(devicetree_api, test_interrupt_controller)
 	/* DT_INST_IRQ_INTC */
 	zassert_true(DT_SAME_NODE(DT_INST_IRQ_INTC(0), TEST_INTC), "");
+
+#ifdef CONFIG_2ND_LEVEL_INTERRUPTS
+	/* the following asserts check if interrupt IDs are encoded
+	 * properly when dealing with a node that consumes interrupts
+	 * from L2 aggregators extending different L1 interrupts.
+	 */
+	zassert_true(DT_IRQN_BY_IDX(TEST_IRQ_EXT, 0) ==
+		     (((70 + 1) << CONFIG_1ST_LEVEL_INTERRUPT_BITS) | 11), "");
+	zassert_true(DT_IRQN_BY_IDX(TEST_IRQ_EXT, 2) ==
+		     (((42 + 1) << CONFIG_1ST_LEVEL_INTERRUPT_BITS) | 12), "");
+#endif /* CONFIG_2ND_LEVEL_INTERRUPTS */
```
might be better to have this under the `test_irq` test case above:

```c
#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT vnd_interrupt_holder_extended
ZTEST(devicetree_api, test_irq_extended)
{
	/* DT_INST_IRQN_BY_IDX */
#ifndef CONFIG_MULTI_LEVEL_INTERRUPTS
	zassert_equal(DT_INST_IRQN_BY_IDX(0, 0), 70, "");
	/* idx 1 is a gpio-controller */
	zassert_equal(DT_INST_IRQN_BY_IDX(0, 2), 50, "");
#else
	zassert_equal(DT_INST_IRQN_BY_IDX(0, 0),
		      ((70 + 1) << CONFIG_1ST_LEVEL_INTERRUPT_BITS) | 11, "");
	/* idx 1 is a gpio-controller */
	zassert_equal(DT_INST_IRQN_BY_IDX(0, 2),
		      ((50 + 1) << CONFIG_1ST_LEVEL_INTERRUPT_BITS) | 12, "");
#endif
}
```

LaurentiuM1234 replied:
moved

cfriedt reviewed (dismissed):
LGTM

ycsin commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -750,6 +750,17 @@ ZTEST(devicetree_api, test_irq)
 	zassert_true(DT_INST_IRQ_HAS_NAME(0, done), "");
 	zassert_false(DT_INST_IRQ_HAS_NAME(0, alpha), "");
+
+#ifdef CONFIG_MULTI_LEVEL_INTERRUPTS
```
we should also make sure that the macro works as expected on such node for non-`CONFIG_MULTI_LEVEL_INTERRUPTS` build

LaurentiuM1234 replied:
fixed

ycsin reviewed (approved):
LGTM, thanks.
Could you please create a bug report for this so that it can be backported?

LaurentiuM1234 commented:
> LGTM, thanks. Could you please create a bug report for this so that it can be backported?

created https://github.com/zephyrproject-rtos/zephyr/issues/69820
